### PR TITLE
HIVE-27344: Add a null check in RecordReaderImpl.close()

### PR DESCRIPTION
**Pull request by ngsg**
### What changes were proposed in this pull request?

Add a null check in RecordReaderImpl.close().

### Why are the changes needed?

Current implementation can throw NPE because the constructor of its super class may call this.close(), which mutates the uninitialized field 'batch'.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

I tested this patch in my local environment.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/RecordReaderImpl.java [hunk 1]
```diff
@@ -165,7 +165,9 @@ public boolean nextBatch(VectorizedRowBatch theirBatch) throws IOException {
   public void close() throws IOException {
     super.close();
     // free the memory for the column vectors
-    batch.cols = null;
+    if (batch != null) {
```
Comment (aturoczy) [inline ql/src/java/org/apache/hadoop/hive/ql/io/orc/RecordReaderImpl.java:168]: It is fine. Personally I like more the following syntax, it is bit more readable to me. (You do **not** need to change) 
```
if(batch == null)
   return;
batch.cols = null
```

**Review by aturoczy — APPROVED** [review]
LGTM

**Comment by deniskuzZ** [comment]
@ngsg, could you please add the testcase that reproduces mentioned issue?

**Comment by ngsg** [comment]
@deniskuzZ  It is hard to add a test case for reproducing the issue because it deeply depends on ORC implementation, especially on the constructor ORC's RecordReaderImpl. The original situation we faced also depends on HADOOP-16109 issue, so it won't be appear with Hadoop releases on which HADOOP-16109 patch is applied. As this issue deeply depends on the outside of Hive and the patch is just a kind of defensive programming, I suggest not adding a test case to reproduce the problem.

**Review by deniskuzZ — APPROVED** [review]
+1, pending tests

**Pull request merged**